File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr) {
            return nullptr;
        }

        ListNode* dummy = new ListNode(0);
        dummy->next = head;

        ListNode* temp = head;
        ListNode* prev = dummy;
        while (temp != nullptr) {
            ListNode* curr = temp;
            temp = temp->next;
            curr->next = prev;
            prev = curr;
        }

        head->next = nullptr;

        return prev;
        
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode h;
    public ListNode reverseList(ListNode head) {
        if (head == null) return head;

        reverse(head);
        head = h;
        return head;
    }

    private void reverse(ListNode curr) {
        if (curr.next == null) {
            h = curr;
            return;
        }

        reverse(curr.next);

        curr.next.next = curr;
        curr.next = null;
    }
}